substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs

```diff
@@ -121,66 +160,107 @@ where
 	}
 }
 
-impl<Address, Call, Signature, Extra: SignedExtension>
-	UncheckedExtrinsic<Address, Call, Signature, Extra>
-{
-	/// New instance of a signed extrinsic aka "transaction".
-	pub fn new_signed(function: Call, signed: Address, signature: Signature, extra: Extra) -> Self {
-		Self { signature: Some((signed, signature, extra)), function }
+impl<Address, Call, Signature, Extension> UncheckedExtrinsic<Address, Call, Signature, Extension> {
+	/// New instance of a bare (ne unsigned) extrinsic. This could be used for an inherent or an
+	/// old-school "unsigned transaction" (which are new being deprecated in favour of general
+	/// transactions).
+	#[deprecated = "Use new_bare instead"]
+	pub fn new_unsigned(function: Call) -> Self {
+		Self::new_bare(function)
 	}
 
-	/// New instance of an unsigned extrinsic aka "inherent".
-	pub fn new_unsigned(function: Call) -> Self {
-		Self { signature: None, function }
+	/// TODO: docs
+	pub fn is_inherent(&self) -> bool {
+		matches!(self.preamble, Preamble::Bare)
+	}
+
+	/// TODO: docs
+	pub fn from_parts(function: Call, preamble: Preamble<Address, Signature, Extension>) -> Self {
+		Self { preamble, function }
+	}
+
+	/// New instance of a bare (ne unsigned) extrinsic.
+	pub fn new_bare(function: Call) -> Self {
+		Self { preamble: Preamble::Bare, function }
+	}
+
+	/// New instance of an old-school signed transaction.
+	pub fn new_signed(
+		function: Call,
+		signed: Address,
+		signature: Signature,
+		tx_ext: Extension,
+	) -> Self {
+		Self { preamble: Preamble::Signed(signed, signature, tx_ext), function }
+	}
+
+	/// New instance of an new-school unsigned transaction.
+	pub fn new_transaction(function: Call, tx_ext: Extension) -> Self {
+		Self { preamble: Preamble::General(tx_ext), function }
 	}
 }
 
-impl<Address: TypeInfo, Call: TypeInfo, Signature: TypeInfo, Extra: SignedExtension + TypeInfo>
-	Extrinsic for UncheckedExtrinsic<Address, Call, Signature, Extra>
+// TODO: We can get rid of this trait and just use UncheckedExtrinsic directly.
```

Comment:
Are you sure? I suppose there is some client side code that made assumptions on this trait, similar to `trait Header` and `trait Block`. If not, this can already be removed as a separate PR in advance, possibly reducing the diff here.